File: python_try/trailerlib.py
# Trailer path planning library
# author: Atsushi Sakai(@Atsushi_twi)

import numpy as np
import matplotlib.pyplot as plt
from scipy.spatial import KDTree

# Vehicle parameters
WB = 3.7  # [m] wheel base: rear to front steer
LT = 8.0  # [m] rear to trailer wheel
W = 2.6   # [m] width of vehicle
LF = 4.5  # [m] distance from rear to vehicle front end of vehicle
LB = 1.0  # [m] distance from rear to vehicle back end of vehicle
LTF = 1.0  # [m] distance from rear to vehicle front end of trailer
LTB = 9.0  # [m] distance from rear to vehicle back end of trailer
MAX_STEER = 0.6  # [rad] maximum steering angle 
TR = 0.5  # Tyre radius [m] for plot
TW = 1.0  # Tyre width [m] for plot

# For collision check
WBUBBLE_DIST = 3.5  # Distance from rear to center of whole bubble
WBUBBLE_R = 10.0  # Whole bubble radius
B = 4.45  # Distance from rear to vehicle back end
C = 11.54  # Distance from rear to vehicle front end
I = 8.55  # Width of vehicle
VRX = np.array([C, C, -B, -B, C])
VRY = np.array([-I / 2.0, I / 2.0, I / 2.0, -I / 2.0, -I / 2.0])


def check_collision(x, y, yaw, kdtree, ox, oy, wbd, wbr, vrx, vry):
    for ix, iy, iyaw in zip(x, y, yaw):
        cx = ix + wbd * np.cos(iyaw)
        cy = iy + wbd * np.sin(iyaw)

        # Whole bubble check
        ids = kdtree.query_ball_point([cx, cy], wbr)
        if len(ids) == 0:
            continue

        # if not rect_check(ix, iy, iyaw, ox[ids], oy[ids], vrx, vry):
        #     return False  # Collision

    return True  # OK


def rect_check(ix, iy, iyaw, ox, oy, vrx, vry):
    c = np.cos(-iyaw)
    s = np.sin(-iyaw)

    for iox, ioy in zip(ox, oy):
        tx = iox - ix
        ty = ioy - iy
        lx = (c * tx - s * ty)
        ly = (s * tx + c * ty)

        sumangle = 0.0
        for i in range(len(vrx) - 1):
            x1 = vrx[i] - lx
            y1 = vry[i] - ly
            x2 = vrx[i + 1] - lx
            y2 = vry[i + 1] - ly
            d1 = np.hypot(x1, y1)
            d2 = np.hypot(x2, y2)
            theta1 = np.arctan2(y1, x1)
            tty = (-np.sin(theta1) * x2 + np.cos(theta1) * y2)
            tmp = (x1 * x2 + y1 * y2) / (d1 * d2)

            tmp = np.clip(tmp, -1.0, 1.0)

            if tty >= 0.0:
                sumangle += np.arccos(tmp)
            else:
                sumangle -= np.arccos(tmp)

        if abs(sumangle) >= np.pi:
            return False  # Collision

    return True  # OK


def calc_trailer_yaw_from_xyyaw(x, y, yaw, init_tyaw, steps):
    """
    Calculate trailer yaw from x, y, yaw lists
    """
    tyaw = np.zeros(len(x))
    tyaw[0] = init_tyaw

    for i in range(1, len(x)):
        tyaw[i] += tyaw[i - 1] + steps[i - 1] / LT * np.sin(yaw[i - 1] - tyaw[i - 1])

    return tyaw


def trailer_motion_model(x, y, yaw0, yaw1, D, d, L, delta):
    """
    Motion model for trailer 
    see: http://planning.cs.uiuc.edu/node661.html#77556
    """
    x += D * np.cos(yaw0)
    y += D * np.sin(yaw0)
    yaw0 += D / L * np.tan(delta)
    yaw1 += D / d * np.sin(yaw0 - yaw1)

    return x, y, yaw0, yaw1


def check_trailer_collision(ox, oy, x, y, yaw0, yaw1, kdtree=None):
    """
    Collision check function for trailer
    """
    if kdtree is None:
        kdtree = KDTree(np.vstack((ox, oy)).T)

    vrxt = np.array([LTF, LTF, -LTB, -LTB, LTF])
    vryt = np.array([-W / 2.0, W / 2.0, W / 2.0, -W / 2.0, -W / 2.0])

    # Bubble parameters
    DT = (LTF + LTB) / 2.0 - LTB
    DTR = (LTF + LTB) / 2.0 + 0.3

    # Check trailer
    if not check_collision(x, y, yaw1, kdtree, ox, oy, DT, DTR, vrxt, vryt):
        return False

    vrxf = np.array([LF, LF, -LB, -LB, LF])
    vryf = np.array([-W / 2.0, W / 2.0, W / 2.0, -W / 2.0, -W / 2.0])

    # Bubble parameters
    DF = (LF + LB) / 2.0 - LB
    DFR = (LF + LB) / 2.0 + 0.3

    # Check front trailer
    if not check_collision(x, y, yaw0, kdtree, ox, oy, DF, DFR, vrxf, vryf):
        return False

    return True  # OK


def plot_trailer(x, y, yaw, yaw1, steer):
    truckcolor = "-k"

    LENGTH = LB + LF
    LENGTHt = LTB + LTF

    truckOutLine = np.array([[-LB, (LENGTH - LB), (LENGTH - LB), -LB, -LB],
                             [W / 2, W / 2, -W / 2, -W / 2, W / 2]])

    trailerOutLine = np.array([[-LTB, (LENGTHt - LTB), (LENGTHt - LTB), -LTB, -LTB],
                               [W / 2, W / 2, -W / 2, -W / 2, W / 2]])

    rr_wheel = np.array([[TR, -TR, -TR, TR, TR],
                         [-W / 12.0 + TW, -W / 12.0 + TW, W / 12.0 + TW, W / 12.0 + TW, -W / 12.0 + TW]])

    rl_wheel = np.array([[TR, -TR, -TR, TR, TR],
                         [-W / 12.0 - TW, -W / 12.0 - TW, W / 12.0 - TW, W / 12.0 - TW, -W / 12.0 - TW]])

    fr_wheel = rr_wheel.copy()
    fl_wheel = rl_wheel.copy()

    tr_wheel = rr_wheel.copy()
    tl_wheel = rl_wheel.copy()

    # Rotation matrices
    Rot1 = np.array([[np.cos(yaw), np.sin(yaw)],
                     [-np.sin(yaw), np.cos(yaw)]])
    Rot2 = np.array([[np.cos(steer), np.sin(steer)],
                     [-np.sin(steer), np.cos(steer)]])
    Rot3 = np.array([[np.cos(yaw1), np.sin(yaw1)],
                     [-np.sin(yaw1), np.cos(yaw1)]])

    # Front wheels transformation
    fr_wheel = Rot2 @ fr_wheel
    fl_wheel = Rot2 @ fl_wheel
    fr_wheel[0, :] += WB
    fl_wheel[0, :] += WB
    fr_wheel = Rot1 @ fr_wheel
    fl_wheel = Rot1 @ fl_wheel

    # Trailer wheels transformation
    tr_wheel[0, :] -= LT
    tl_wheel[0, :] -= LT
    tr_wheel = Rot3 @ tr_wheel
    tl_wheel = Rot3 @ tl_wheel

    # Transform truck and trailer outlines
    truckOutLine = Rot1 @ truckOutLine
    trailerOutLine = Rot3 @ trailerOutLine

    # Plotting
    plt.plot(trailerOutLine[0, :] + x, trailerOutLine[1, :] + y, truckcolor)
    plt.plot(tr_wheel[0, :] + x, tr_wheel[1, :] + y, truckcolor)
    plt.plot(tl_wheel[0, :] + x, tl_wheel[1, :] + y, truckcolor)

    plt.plot(truckOutLine[0, :] + x, truckOutLine[1, :] + y, truckcolor)
    plt.plot(fr_wheel[0, :] + x, fr_wheel[1, :] + y, truckcolor)
    plt.plot(fl_wheel[0, :] + x, fl_wheel[1, :] + y, truckcolor)
    plt.plot(rr_wheel[0, :] + x, rr_wheel[1, :] + y, truckcolor)
    plt.plot(rl_wheel[0, :] + x, rl_wheel[1, :] + y, truckcolor)


if __name__ == "__main__":
    # Test example
    ox = [i for i in range(50)]
    oy = [i for i in range(50)]
    
    # Create kd-tree for collision checking
    kdtree = KDTree(np.vstack((ox, oy)).T)

    # Test for collision check function
    x = np.array([0.0])
    y = np.array([0.0])
    yaw = np.array([0.0])
    yaw1 = np.array([0.0])

    is_collision = check_trailer_collision(ox, oy, x, y, yaw, yaw1, kdtree)
    print(f"Is collision: {is_collision}")

    # Test plot
    plt.figure()
    plot_trailer(0.0, 0.0, 0.0, 0.0, 0.0)
    plt.axis("equal")
    plt.show()

File: lib_cpp_hybrid_a_star/src/trailerlib.cpp
#include "lib_cpp_hybrid_a_star/trailerlib.hpp"

namespace planning
{
    const std::vector<double> VehicleParams::VRX = {VehicleParams::C, VehicleParams::C
                                    , -VehicleParams::B, -VehicleParams::B, VehicleParams::C };
    const std::vector<double> VehicleParams::VRY = {-VehicleParams::I/2.0, VehicleParams::I/2.0
                        , VehicleParams::I/2.0, -VehicleParams::I/2.0, -VehicleParams::I/2.0};

    TrailerLib::TrailerLib(){

    }

    bool TrailerLib::rect_check(const Eigen::Vector3d& current_pose, const Eigen::MatrixXd& obses
                                , const std::vector<double> vrx, const std::vector<double> vry){

        double c = cos(-current_pose.z());
        double s = sin(-current_pose.z());
        for(int j=0; j< obses.rows(); j++){
            Eigen::VectorXd obs = obses.row(j);
            double tx = obs.x() - current_pose.x();
            double ty = obs.y() - current_pose.y();
            double lx = (c*tx - s*ty);
            double ly = (s*tx + c*ty);
            double sumangle = 0.0;
            
            for(int i=0; i< vrx.size()-1; i++){
                double x1 = vrx[i] - lx;
                double y1 = vry[i] - ly;
                double x2 = vrx[i+1] - lx;
                double y2 = vry[i+1] - ly;
                double d1 = hypot(x1,y1);
                double d2 = hypot(x2,y2);
                double theta1 = atan2(y1,x1);
                double tty = (-sin(theta1)*x2 + cos(theta1)*y2);
                double tmp = (x1*x2+y1*y2)/(d1*d2);
                tmp = std::clamp(tmp, 0.0, 1.0);
                sumangle += (tty >= 0.0) ? acos(tmp) :  -acos(tmp);
            }
            if(sumangle >= M_PI){
                return false;
            }
        }
        return true;
    }

    bool TrailerLib::check_collision(Eigen::MatrixXd& poses, grid_search::KDTree& kd_tree
                        , const Eigen::MatrixXd& obses, double wbd, double wbr
                        , const std::vector<double> vrx, const std::vector<double> vry){
        
        double radius = wbr;
        for(int i=0; i<poses.rows(); i++){
            Eigen::VectorXd next_pose = poses.row(i);
            double cx = next_pose[0] + wbd*cos(next_pose[2]);
            double cy = next_pose[1] + wbd*sin(next_pose[2]);
            Eigen::Vector3d next_estimated_pose = next_pose.head(3);
            std::vector<nanoflann::ResultItem<size_t, double>> indices_dists;
            nanoflann::RadiusResultSet<double, size_t> result_set(radius, indices_dists);
            Eigen::Vector2d query_point(cx, cy);
            kd_tree.findNeighbors(result_set, query_point.data(), nanoflann::SearchParameters());
            if(indices_dists.size() < 1){
                continue;
            }
            Eigen::MatrixXd obs_subset(indices_dists.size(), 2);
            for (size_t i = 0; i < indices_dists.size(); ++i) {
                obs_subset.row(i) << obses(indices_dists[i].first, 0), obses(indices_dists[i].first, 1);
            }
            if (!rect_check(next_estimated_pose, obs_subset, vrx, vry)){
                return false;
            }
        }
        return true;
    }

    bool TrailerLib::calc_trailer_yaw_from_xyyaw(Eigen::MatrixXd& poses, const double init_tyaw
                                                        , Eigen::VectorXd& steps, Eigen::VectorXd& yaws){
        Eigen::VectorXd yaw = poses.col(2);
        yaws.resize(yaw.size());
        yaws.setZero();
        yaws[0] = init_tyaw;
        for(int i=1; i<yaws.size(); i++){
            double steps_d =  steps[i-1];
            double delta_theta = (steps_d/VehicleParams::LT)*sin(yaw[i-1] - yaws[i-1]);
            yaws[i] += yaws[i-1] + delta_theta;
        }
        return true;
    } 

    bool TrailerLib::check_trailer_collision(const Eigen::MatrixXd& obses
                                    , Eigen::MatrixXd& poses, grid_search::KDTree& kd_tree){
        std::vector<double> vrxt = {VehicleParams::LTF, VehicleParams::LTF, -VehicleParams::LTB
                                    , -VehicleParams::LTB, VehicleParams::LTF};
        std::vector<double> vryt = {-VehicleParams::W/2.0, VehicleParams::W/2.0
                                    , VehicleParams::W/2.0, -VehicleParams::W/2.0, -VehicleParams::W/2.0};

        double DT = (VehicleParams::LTF + VehicleParams::LTB)/2.0 - VehicleParams::LTB;
        double DTR = (VehicleParams::LTF + VehicleParams::LTB)/2.0 + 0.3;


        Eigen::MatrixXd back_path = poses;
        Eigen::MatrixXd front_path = poses;
        back_path.col(2) = poses.col(3);

        if (!check_collision(back_path, kd_tree, obses, DT, DTR, vrxt, vryt)){
            return false;
        }

        std::vector<double> vrxf = {VehicleParams::LF, VehicleParams::LF, -VehicleParams::LB, -VehicleParams::LB, VehicleParams::LF};
        std::vector<double> vryf = {-VehicleParams::W/2.0, VehicleParams::W/2.0
                                    , VehicleParams::W/2.0, -VehicleParams::W/2.0, -VehicleParams::W/2.0};
        double DF = (VehicleParams::LF + VehicleParams::LB)/2.0 - VehicleParams::LB;
        double DFR = (VehicleParams::LF + VehicleParams::LB)/2.0 + 0.3;

        if (!check_collision(front_path, kd_tree, obses, DF, DFR, vrxf, vryf)){
            return false;
        }
        return true;
    }

    void TrailerLib::plot_trailer(double x, double y, double yaw, double yaw1, double steer){

        std::string truckcolor = "-k";
        typedef VehicleParams VP;
        double LENGTH = VP::LB+VP::LF;
        double LENGTHt = VP::LTB + VP::LTF;
        Eigen::MatrixXd truckOutLine(2, 5);
        truckOutLine << -VP::LB, (LENGTH - VP::LB), (LENGTH - VP::LB), (-VP::LB), -VP::LB
                        , VP::W/2, VP::W/2, -VP::W/2, -VP::W/2, VP::W/2;
        Eigen::MatrixXd trailerOutLine(2, 5);
        trailerOutLine << -VP::LTB, (LENGTHt - VP::LTB), (LENGTHt - VP::LTB), (-VP::LTB), (-VP::LTB)
                            ,VP::W/2, VP::W/2, -VP::W/2, -VP::W/2, VP::W/2;

        Eigen::MatrixXd rr_wheel(2, 5);
        rr_wheel << VP::TR, -VP::TR, -VP::TR, VP::TR, VP::TR, -VP::W/12.0+VP::TW
                    ,  -VP::W/12.0+VP::TW, VP::W/12.0+VP::TW, VP::W/12.0+VP::TW, -VP::W/12.0+VP::TW;

        Eigen::MatrixXd rl_wheel(2, 5);
        rl_wheel << VP::TR, -VP::TR, -VP::TR, VP::TR, VP::TR, -VP::W/12.0-VP::TW
                    ,  -VP::W/12.0-VP::TW, VP::W/12.0-VP::TW, VP::W/12.0-VP::TW, -VP::W/12.0-VP::TW;
                    
        Eigen::MatrixXd fr_wheel(2, 5);
        fr_wheel << VP::TR, -VP::TR, -VP::TR, VP::TR, VP::TR,
                    -VP::W/12.0+VP::TW,  -VP::W/12.0+VP::TW, VP::W/12.0+VP::TW, VP::W/12.0+VP::TW, -VP::W/12.0+VP::TW;

        Eigen::MatrixXd fl_wheel(2, 5);
        fl_wheel <<  VP::TR, -VP::TR, -VP::TR, VP::TR, VP::TR,
                    -VP::W/12.0-VP::TW,  -VP::W/12.0-VP::TW, VP::W/12.0-VP::TW, VP::W/12.0-VP::TW, -VP::W/12.0-VP::TW; 

        Eigen::MatrixXd tr_wheel(2, 5);
        tr_wheel << VP::TR, -VP::TR, -VP::TR, VP::TR, VP::TR,
                    -VP::W/12.0+VP::TW,  -VP::W/12.0+VP::TW, VP::W/12.0+VP::TW, VP::W/12.0+VP::TW, -VP::W/12.0+VP::TW;
        
        Eigen::MatrixXd tl_wheel(2, 5);
        tl_wheel << VP::TR, -VP::TR, -VP::TR, VP::TR, VP::TR,
                    -VP::W/12.0-VP::TW,  -VP::W/12.0-VP::TW, VP::W/12.0-VP::TW, VP::W/12.0-VP::TW, -VP::W/12.0-VP::TW;
        
        Eigen::MatrixXd Rot1(2, 2);
        Rot1 << cos(yaw), sin(yaw), -sin(yaw), cos(yaw);

        Eigen::MatrixXd Rot2(2, 2);
        Rot2 << cos(steer), sin(steer), -sin(steer), cos(steer);

        Eigen::MatrixXd Rot3(2, 2);
        Rot3 << cos(yaw1), sin(yaw1), -sin(yaw1), cos(yaw1);

        fr_wheel = Rot2*fr_wheel;
        fl_wheel = Rot2*fl_wheel;
        fr_wheel.row(0) = fr_wheel.row(0).array() + VP::WB;
        fl_wheel.row(0) = fl_wheel.row(0).array() + VP::WB;
        fr_wheel = Rot1.transpose()*fr_wheel;
        fl_wheel = Rot1.transpose()*fl_wheel;
        tr_wheel.row(0) = tr_wheel.row(0).array() - VP::LT;
        tl_wheel.row(0) = tl_wheel.row(0).array() - VP::LT;
        tr_wheel = Rot3.transpose()*tr_wheel;
        tl_wheel = Rot3.transpose()*tl_wheel;

        truckOutLine = Rot1.transpose()*truckOutLine;
        trailerOutLine = Rot3.transpose()*trailerOutLine;

        rr_wheel = Rot1.transpose()*rr_wheel;
        rl_wheel = Rot1.transpose()*rl_wheel;

        truckOutLine.row(0) = truckOutLine.row(0).array() + x;
        truckOutLine.row(1) = truckOutLine.row(1).array() + y;
        trailerOutLine.row(0) = trailerOutLine.row(0).array() + x;
        trailerOutLine.row(1) = trailerOutLine.row(1).array() + y;

        fr_wheel.row(0) = fr_wheel.row(0).array() + x;
        fr_wheel.row(1) = fr_wheel.row(1).array() + y;
        rr_wheel.row(0) = rr_wheel.row(0).array() + x;
        rr_wheel.row(1) = rr_wheel.row(1).array() + y;
        fl_wheel.row(0) = fl_wheel.row(0).array() + x;
        fl_wheel.row(1) = fl_wheel.row(1).array() + y;
        rl_wheel.row(0) = rl_wheel.row(0).array() + x;
        rl_wheel.row(1) = rl_wheel.row(1).array() + y;

        tr_wheel.row(0) = tr_wheel.row(0).array() + x;
        tr_wheel.row(1) = tr_wheel.row(1).array() + y;
        tl_wheel.row(0) = tl_wheel.row(0).array() + x;
        tl_wheel.row(1) = tl_wheel.row(1).array() + y;


        auto plotOutline = [&](const Eigen::MatrixXd& outline) {
            std::vector<double> vec_x(outline.cols()), vec_y(outline.cols());
            for (int i = 0; i < outline.cols(); ++i) {
                vec_x[i] = outline(0, i);
                vec_y[i] = outline(1, i);
            }
            plt::plot(vec_x, vec_y, truckcolor);
        };

        // plt::figure();
        plotOutline(truckOutLine);
        plotOutline(trailerOutLine);
        plotOutline(fr_wheel);
        plotOutline(rr_wheel);
        plotOutline(rl_wheel);
        plotOutline(fl_wheel);
        plotOutline(tr_wheel);
        plotOutline(tl_wheel);
        plt::plot({x}, {y}, "*");
        // plt::axis("equal");
        // plt::show();

    }
}

// int main(int argc, char *argv[]) {

//     double x = 0.0;
//     double y = 0.0;
//     double yaw0 = math_utility::deg2rad(10.0);
//     double yaw1 = math_utility::deg2rad(-10.0);
//     planning::TrailerLib trailer_lib;
//     trailer_lib.plot_trailer(x, y, yaw0, yaw1, 0.0);
// }

